File: 1680-concatenation-of-consecutive-binary-numbers/1680-concatenation-of-consecutive-binary-numbers.java
class Solution {
    public int concatenatedBinary(int n) {
        long result = 0;
        long mod = (int) 1e9 + 7;
        long currentPowerOfTwo = 1;
        int numberOfBits = 0;
        
        for (int i = 1; i <= n; i++) {
            if (i == currentPowerOfTwo) {
                numberOfBits++;
            }
            
            result = ((result << numberOfBits) + i) % mod;
            
            if (i == currentPowerOfTwo) {
                currentPowerOfTwo *= 2;
            }
        }
        
        return (int) result;
    }
}

File: 1680-concatenation-of-consecutive-binary-numbers/1680-concatenation-of-consecutive-binary-numbers.cpp
class Solution {
public:
    int concatenatedBinary(int n) {
        long result = 1L;
        int numberOfBits = 1;
        
        for (int i = 2; i <= n; i++) {
            if (i == (1 << numberOfBits)) {
                numberOfBits++;
            }
            
            result = (result << numberOfBits) + i;
            
            if (result >= 1000000007) {
                result = result - (result / 1000000007) * 1000000007;
            }
        }
        
        return result;
    }
};